cmd/calico-ipam-upgrade/main.go

```diff
@@ -0,0 +1,193 @@
+)
+
+var (
+	binariesToDisable = []string{
+		"/host/opt/cni/bin/calico",
+		"/host/opt/cni/bin/host-local",
+	}
+)
+
+func main() {
+	// Set up logging formatting.
+	log.SetFormatter(&logutils.Formatter{})
+
+	// Install a hook that adds file/line no information.
+	log.AddHook(&logutils.ContextHook{})
+
+	// Check to see if the system is still using host-local
+	// by checking the existence of the path.
+	var err error
+	if _, err = os.Stat(ipAllocPath); err != nil && os.IsNotExist(err) {
+		log.Info("no migration necessary, exiting...")
+		os.Exit(0)
+	}
+
+	// First, disable the necessary binaries so it can't be used.
+	for _, binary := range binariesToDisable {
+		if err = toggleFile(binary, true); err != nil {
+			log.WithField("file", binary).WithError(err).Fatal("failed to disable binary")
+		}
+	}
+	defer func() {
+		if err != nil {
+			rollback()
+		}
+	}()
+	<-time.After(time.Second)
+
+	// Initialize necessary variables
+	nodename := os.Getenv("KUBERNETES_NODE_NAME")
+	cfg, err := apiconfig.LoadClientConfig("")
+	if err != nil {
+		log.WithError(err).Fatal("failed to load api client config")
+	}
+	cfg.Spec.DatastoreType = apiconfig.Kubernetes
+	calicoClient, err := client.New(*cfg)
+	if err != nil {
+		log.WithError(err).Fatal("failed to initialize api client")
+	}
+
+	// Disable auto-allocation of blocks
+	if err = calicoClient.IPAM().SetIPAMConfig(context.Background(), ipam.IPAMConfig{
+		AutoAllocateBlocks: false,
+		StrictAffinity:     true,
+	}); err != nil {
+		log.WithError(err).Fatal("failed to disable auto-allocation of block")
+	}
+
+	// Read in all the files in the host-local directory.
+	files, err := ioutil.ReadDir(ipAllocPath)
+	if err != nil {
+		log.WithField("path", ipAllocPath).WithError(err).Fatal("failed to read path")
+	}
+
+	// For each file, convert it into an IP allocation and then delete the file.
+	for _, f := range files {
+		fname := path.Join(ipAllocPath, f.Name())
+
+		// Delete and skip the last reserved IP.
+		if strings.Contains(f.Name(), "last") {
+			if err := os.Remove(fname); err != nil {
+				log.WithField("file", fname).WithError(err).Fatal("failed to remove file")
+			}
+			continue
+		}
+
+		// The name of the file is its IP address.
+		ip, _, err := cnet.ParseCIDR(fmt.Sprintf("%s/32", f.Name()))
+		if err != nil {
+			log.WithField("file", f.Name()).WithError(err).Fatal("failed to parse CIDR")
+		}
+
+		// The contents are the container ID.
+		b, err := ioutil.ReadFile(fname)
+		if err != nil {
+			log.WithField("file", fname).WithError(err).Fatal("failed to read file")
+		}
+
+		// Store allocation to datastore
+		err = assignCalicoIP(calicoClient, *ip, string(b), nodename)
+		if err != nil {
+			log.WithField("nodename", nodename).WithError(err).Fatal("failed to assign IP to calico backend")
+		}
+
+		// Delete the file
+		if err := os.Remove(fname); err != nil {
+			log.WithField("file", fname).WithError(err).Fatal("failed to remove file")
```

Comment:
I think rather than exit the process on each of these error conditions, we should implement some retry logic.

We could wrap this in another function which orchestrates the retries, e.g.:

```go
  for i:=0; i< numretries; i++ {
    if err := doUpgrade(); err != nil { continue}
    break
  }
```